azure-iot-device/azure/iot/device/patch.py

```diff
@@ -0,0 +1,133 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module provides patches used to dynamically modify items from the libraries"""
+
+import sys
+import inspect
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: make this work for Python 2.7 and 3.4
+def add_shims_for_inherited_methods(target_class):
+    """Dynamically add overriding, pass-through shim methods for all public inherited methods
+    on a child class, which simply call into the parent class implementation of the same method.
+
+    These shim methods will include the same docstrings as the method from the parent class.
+
+    This currently only works for Python 3.5+
+
+    :param target_class: The child class to add shim methods to
+    """
+
+    # Depending on how the method was defined, it could be either a function or a method.
+    # Thus we need to find the union of the two sets.
+    # Here instance methods are considered functions because they are not yet bound to an instance
+    # of the class. Classmethods on the other hand, are already bound, and show up as methods.
+    # It also is worth noting that async functions/methods ARE picked up by this introspection.
+    class_functions = inspect.getmembers(target_class, predicate=inspect.isfunction)
+    class_methods = inspect.getmembers(target_class, predicate=inspect.ismethod)
+    all_methods = class_functions + class_methods
+
+    # This list of attributes gives us a lot of information, but we only are using it to get
+    # the defining class of a given method.
+    class_attributes = inspect.classify_class_attrs(target_class)
+
+    # We must alias classnames to prevent naming collisions when this fn is called multiple times
+    # with classes that share a name. If we've already used this classname, add trailing underscore(s)
+    classname_alias = target_class.__name__
+    while classname_alias in globals():
+        classname_alias += "_"
+
+    # Import the class we're adding methods to, so that functions defined in this scope can use super()
+    class_module = inspect.getmodule(target_class)
+    exec(
+        "from {module} import {target_class} as {alias}".format(
+            module=class_module.__name__, target_class=target_class.__name__, alias=classname_alias
+        ),
+        globals(),
+    )
+
+    for method in all_methods:
+        method_name = method[0]
+        method_obj = method[1]
+        # We can index on 0 here because the list comprehension will always be exactly 1 element
+        method_attribute = [att for att in class_attributes if att.name == method_name][0]
```

Comment:
Nice trick.  makes me wonder if there's a clearer way to do this.

Reply:
Too bad there's not a builtin for this.  I've also found next(att for att in class_attributes if att.name == mathod_name)) but that doesn't add any clarity as far as I'm concerned

---
In reply to: [315324838](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315324838) [](ancestors = 315324838)

Reply:
Yeah, I tried to find something more elegant, but unfortunately, there doesn't seem to be anything. I think the core problem is that there's no real way to guarantee that no more than one item in the list will meet the select condition.

The only alternative I can think of involves a for loop that checks the condition issuing a `break` after the correct attribute is found.

```python
method_attribute = None
for att in class_attributes:
    if att.name == method_name:
        method_attribute = att
        break
```

This is I suppose, more correct, but takes up more lines. Perhaps it's clearer?

---
In reply to: [315326449](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315326449) [](ancestors = 315326449,315324838)

Reply:
I think your code is just fine.  I'm just lamenting the absence of a better solution because this is a common problem. 

---
In reply to: [315340955](https://github.com/Azure/azure-iot-sdk-python-preview/pull/164#discussion_r315340955) [](ancestors = 315340955,315326449,315324838)